stan/math/prim/fun/select.hpp

```diff
@@ -0,0 +1,173 @@
+ * @param c Boolean condition value.
+ * @param y_true Value to return if condition is true.
+ * @param y_false Value to return if condition is false.
+ */
+template <typename T_true, typename T_false,
+          require_all_stan_scalar_t<T_true, T_false>* = nullptr>
+inline auto select(const bool c, const T_true y_true, const T_false y_false) {
+  return c ? y_true : y_false;
+}
+
+/**
+ * Return the second argument if the first argument is true
+ * and otherwise return the third argument. Eigen expressions are
+ * evaluated so that the return type is the same for both branches.
+ *
+ * Both containers must have the same plain type. The scalar type
+ * of the return is determined by the return_type_t<> type trait.
+ *
+ * Overload for use with two containers.
+ *
+ * @tparam T_true type of the true argument
+ * @tparam T_false type of the false argument
+ * @param c Boolean condition value.
+ * @param y_true Value to return if condition is true.
+ * @param y_false Value to return if condition is false.
+ */
+template <
+    typename T_true, typename T_false,
+    typename T_return = return_type_t<T_true, T_false>,
+    typename T_true_plain = promote_scalar_t<T_return, plain_type_t<T_true>>,
+    typename T_false_plain = promote_scalar_t<T_return, plain_type_t<T_false>>,
+    require_all_container_t<T_true, T_false>* = nullptr,
+    require_all_same_t<T_true_plain, T_false_plain>* = nullptr>
+inline T_true_plain select(const bool c, const T_true y_true,
+                           const T_false y_false) {
+  check_matching_dims("select", "y_true", y_true, "y_false", y_false);
+  if (c) {
+    return y_true;
+  } else {
+    return y_false;
+  }
+}
+
+/**
+ * Return the second Eigen argument if the first argument is true
+ * and otherwise return the second Eigen argument. Overload for use with one
+ * scalar and one Eigen object.
+ *
+ * If chosen, the scalar is returned as an Eigen object of the same size and
+ * plain type as the provided argument. Consequently, any Eigen expressions are
+ * evaluated so that the return type is the same for both branches.
+ *
+ * @tparam T_true type of the true argument
+ * @tparam T_false type of the false argument
+ * @param c Boolean condition value.
+ * @param y_true Value to return if condition is true.
+ * @param y_false Value to return if condition is false.
+ */
+template <typename T_true, typename T_false,
+          typename ReturnT = promote_scalar_t<return_type_t<T_true, T_false>,
+                                              plain_type_t<T_true>>,
+          require_container_t<T_true>* = nullptr,
+          require_stan_scalar_t<T_false>* = nullptr>
+inline ReturnT select(const bool c, const T_true& y_true,
+                      const T_false& y_false) {
+  if (c) {
+    return y_true;
+  } else {
+    return apply_scalar_binary(
+        y_true, y_false,
+        [](const auto& y_true_inner, const auto& y_false_inner) {
+          return y_false_inner;
+        });
```

Comment:
The confusion here was probably due to the docs. I had written that the overloads were for a Eigen + scalar combination, but they're actually for general container + scalar combinations. So the `apply_scalar_binary` here allows this to work with arbitrarily nested containers. For example:
```cpp
select(bool, std::vector<std::vector<Eigen::MatrixXd>>, double)
```

Reply:
I'm still kind of confused. If `c` is false then we go the the `apply_scalar_binary` branch which then returns `y_false_inner`. I'm not really following how a double would be promoted to a std::vector<std::vector<Eigen::MatrixXd>> in your code above?

```c++
double dbl = 3.4;
std::vector<std::vector<Eigen::MatrixXd>> stdvec = // ...
std::vector<std::vector<Eigen::MatrixXd>> result = select(true, stdvec, dbl);
```

Reply:
Can you add a test for `std::vector<std::vector<Eigen::MatrixXd>>` and `double`?

Reply:
Because the return type of the `apply_scalar_binary` call will be the container that is passed, and the functor is specifying that every element of the container is the `y_false` value (which is a scalar in this overload)

Also added a test of these different combinations so you can see the behaviour in action as well